libtenzir/builtins/operators/load_balance.cpp

```diff
@@ -0,0 +1,396 @@
+
+  auto write(table_slice events) -> caf::result<void> {
+    TENZIR_ASSERT(events.rows() > 0);
+    if (not reads.empty()) {
+      TENZIR_WARN("writing {} events directly", events.rows());
+      reads.front().deliver(std::move(events));
+      reads.pop_front();
+      return {};
+    }
+    TENZIR_WARN("writing {} events delayed", events.rows());
+    writes.emplace_back(std::move(events), self->make_response_promise<void>());
+    return writes.back().second;
+  }
+
+  auto read() -> caf::result<table_slice> {
+    if (not writes.empty()) {
+      auto events = std::move(writes.front().first);
+      TENZIR_WARN("reading {} events directly", events.rows());
+      writes.front().second.deliver();
+      writes.pop_front();
+      return events;
+    }
+    if (finished) {
+      return table_slice{};
+    }
+    TENZIR_WARN("reading events delayed");
+    reads.emplace_back(self->make_response_promise<table_slice>());
+    return reads.back();
+  }
+
+  void finish() {
+    if (finished) {
+      return;
+    }
+    finished = true;
+    TENZIR_WARN("load_balancer finished and marks {} outstanding reads as "
+                "done",
+                reads.size());
+    // If there are any outstanding reads, we know that there are no remaining
+    // writes. Thus it's fine to mark all outstanding reads as done.
+    for (auto& read : reads) {
+      read.deliver(table_slice{});
+    }
+    reads.clear();
+  }
+};
+
+class load_balance_source final : public crtp_operator<load_balance_source> {
+public:
+  load_balance_source() = default;
+
+  explicit load_balance_source(load_balancer_actor load_balancer)
+    : load_balancer_{std::move(load_balancer)} {
+  }
+
+  auto name() const -> std::string override {
+    return "internal-load-balance-source";
+  }
+
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    TENZIR_WARN("beginning execution of load_balance_source");
+    TENZIR_ASSERT(load_balancer_);
+    while (true) {
+      auto result = table_slice{};
+      ctrl.set_waiting(true);
+      ctrl.self()
+        .request(load_balancer_, caf::infinite, atom::read_v)
+        .then(
+          [&](table_slice slice) {
+            result = std::move(slice);
+            ctrl.set_waiting(false);
+          },
+          [&](const caf::error& err) {
+            // This should never happen.
+            diagnostic::error("load balancer read failed: {}", err)
+              .emit(ctrl.diagnostics());
+          });
+      co_yield {};
+      // We signal completion with an empty table slice.
+      if (result.rows() == 0) {
+        TENZIR_WARN("load_balance_source detected end");
+        break;
+      }
+      TENZIR_WARN("load_balance_source read {} events", result.rows());
+      co_yield std::move(result);
+    }
+  }
+
+  auto optimize(const expression& filter, event_order order) const
+    -> optimize_result override {
+    TENZIR_UNUSED(filter, order);
+    return do_not_optimize(*this);
+  }
```

Comment:
Since `load_balance` makes no guarantees about the order, I think we can return `event_order::unordered` here.

(Technically only if there are at least two load-balanced pipelines.)

Reply:
I think we can do it either way. With the current setup, we guarantee (without having documented it) that events that arrive at a single destination are ordered. This is no longer true when we return unordered here, so the guarantees you get a strictly weaker. 

It's fine if that's what we want to do, but then I'd argue it's also the case if there is only a single pipeline. Either the operator is allowed to reorder events to a single destination, or it's not.

What do you want? I'm fine with either.

(cc @IyeOnline, who also raised that point)

Reply:
Given that there are significant performance improvements for some operators when they're unordered and that you primarily use load balancing when you care about performance I would argue that we should make it unordered unconditionally.